### PR TITLE
Adding support for optional list to int

### PR DESCRIPTION
**Pull request by nikithamalgifb**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53998 Adding support for optional list to int**

Fixes issue: #48408

Test:
-------
python test/test_jit.py -k test_optional_inputs

### CODE REVIEW

File: torch/csrc/jit/frontend/schema_matching.cpp [hunk 1]
```diff
@@ -20,6 +20,15 @@ static inline bool isIntOrFloatUsedAsList(
     const Argument& arg) {
   // Look for int[N] or float[N]
   const auto& v_type = value->type();
+  // Check if value is optional Type of Int or Float
+  if (value->type()->kind() == OptionalType::Kind) {
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/schema_matching.cpp:24]: Most of logic here is identical to non-optional case. Could we simplify the implementation a little bit to something like this?
```
if (value->type()->kind() == OptionalType::Kind) {
    v_type = unwrapOptional(v_type);  // You may need to remove `const` modifier of `v_type`.
}

... original logic stay as is ...
```

File: test/test_jit.py [hunk 2]
```diff
@@ -1085,6 +1085,13 @@ def fn(*args):
                 self.assertFalse(fn.has_trace_for(*unk_config))
         self.assertEqual(fn.hits, 0)
 
+    def test_optional_inputs(self):
+        def fn(t, n: Optional[int]):
+            return torch.fft.fftn(t, n)
+
+            A = torch.rand(2, 3)
```
Comment (gmagogsfm) [inline test/test_jit.py:1092]:
```suggestion
            a = torch.rand(2, 3)
```

**Review by gmagogsfm — COMMENTED** [review]
Thanks for fixing it! Just a few minor code clean ups needed before we can land it. Also, do you mind improving PR title and description to provide more context?

File: torch/csrc/jit/frontend/schema_matching.cpp (hunk 1 above)
Comment (nikithamalgifb) [reply torch/csrc/jit/frontend/schema_matching.cpp:24]: Agreed and fixed.

File: test/test_jit.py (hunk 2 above)
Comment (nikithamalgifb) [reply test/test_jit.py]: Fixed

File: torch/csrc/jit/frontend/schema_matching.cpp [hunk 3]
```diff
@@ -148,6 +152,11 @@ static Value* tryMatchArgument(
   // also allow single ints/floats to be passed in their place. The single
   // int/float is then repeated to the length of the list
   if (isIntOrFloatUsedAsList(value, arg)) {
+    if (value->type()->kind() == OptionalType::Kind) {
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/schema_matching.cpp:155]: This feels slightly out of place. Basically the implementation of `isIntOrFloatUsedAsList` "leaked" outside. Maybe we should change behavior of `isIntOrFloatUsedAsList` to `maybeUseIntOrFloatAsList`, which returns an updated value that includes logic that are currently in line 155 to line 162 (even the list creation part).

**Review by gmagogsfm — COMMENTED** [review]
Thanks for fixing it! Looking pretty good now. Just one minor nit pick. Also, could you update PR description and title to be more descriptive?

File: torch/csrc/jit/frontend/schema_matching.cpp (hunk 3 above)
Comment (nikithamalgifb) [reply torch/csrc/jit/frontend/schema_matching.cpp]: Now changed to `maybeUseIntOrFloatAsList`

**Pull request closed without merging**